### ISSUE TITLE
Word prediction known issues / crashes

### ISSUE THREAD

**Issue by patrickgold** [florisboard/florisboard#1268]
This issue is a collector for the numerous issues and crashes word prediction in 0.3.13 causes. Word prediction is scheduled to be completely reworked in 0.3.15/0.3.16.

If you experience any of the below listed bullet points or on [this issue collection](https://github.com/florisboard/florisboard/issues?q=is%3Aissue+is%3Aopen+label%3A%22area%3A+word-prediction%22), **DO NOT FILE A SEPARATE ISSUE**! These issues are already known and a separate issue only clutters the issue list even more. New word prediction issues will be closed as a duplicate of this issue here.

Word prediction in 0.3.13 and 0.3.14:
- Are English (US) only
- May not load even if enabled if the Java heap size is too small, in this case you just do not have word predictions atm
- If the dictionary loads it may cause various OOM (OutOfMemory) crashes, like in #677. In this case disable word predictions and restart FlorisBoard
- Suggested words are always lowercase, the caps/shift key is ignored
- Suggested words do not suggest good words for some inputs
- ...

**Comment by Glitchy-Tozier** [comment]
@patrickgold Looks good to me. The only thing you could do is move the `DO NOT CRETE DUPLICATES` in front of the list.

**Comment by patrickgold** [comment]
> The only thing you could do is move the DO NOT CRETE DUPLICATES in front of the list.

Fixed it, thanks for your feedback!

**Comment by Glitchy-Tozier** [comment]
@patrickgold Should we close issues as duplicates of this one once they pop up? (I'm asking as this seems to be more of a collection than a one-problem-issue)

**Comment by patrickgold** [comment]
Issues which are related to things like:

- suggestions don't work in latest version
- OOM crash logs when suggestions=true
- suggestions ignore capitalization
- suggestions give back weird unsable words (they do but atm there's nothing to do about it).

etc.etc. should be closed as a duplicate of this issue. So basically stuff we have heard a lot of times already and which are planned to be addressed in 0.3.15.

**Comment by abda11ah** [comment]
Hi,
Why don't you import word predictions functions from openboard. Their word predictions are perfect and works even with french language.
https://github.com/dslul/openboard

I would like to switch to florisboard but I can't because of the lack of french word prediction.

**Comment by Glitchy-Tozier** [comment]
@abda11ah 
1. The issue isn't the difficulty of implementing word-suggestions. There already exists a way better version of word-suggestions, but we temporarily had to regress to the old system and delay the improved suggestions for ~2 versions.
2. Unfortunately, OpenBoard's word-predictions aren't perfect either. For example, the keyboard doesn't properly learn your words/phrases, while you type.

**Comment by sabzo** [comment]
@Glitchy-Tozier @abda11ah OpenBoard does actually learn word-predictions as you do type. They use a unigram/binary-gram technique, and on the fly they create a user generated "dictionary". This user generated dictionary works in conjunction with the wordlist dictionaries, however the learning seems to be happening in the user-generated dictionary.  The c-code is all over the place and not easy to read

**Comment by Glitchy-Tozier** [comment]
@sabzo If this is the case, I wasn't able to notice any improvements over a 1-2 week period, even with repeatedly typed phrases. (Like the ending of an email)

**Comment by ahmed-tasaly** [comment]
In Arabic language (layout)
Suggest shows multiple inputs 
English and Arabic together

**Comment by heinhuiz** [comment]
Is it possible to by default include a space character when a suggested word is selected?

**Comment by patrickgold** [comment]
> Is it possible to by default include a space character when a suggested word is selected?

Normally a "phantom space is inserted" after selecting a suggested word, which means a space character will appear if you type any letter (not with symbols though, in this case the phantom space is discarded).

**Comment by heinhuiz** [comment]
> > Is it possible to by default include a space character when a suggested word is selected?
> 
> Normally a "phantom space is inserted" after selecting a suggested word, which means a space character will appear if you type any letter (not with symbols though, in this case the phantom space is discarded).

You're right, I've only noticed that after posting this issue. It would however be better if that space is actually visible, so that one knows where and what he is typing. Just move the cursor one position to the right automatically after the word is replaced.

**Comment by heinhuiz** [comment]
> > This issue is a collector for the numerous issues and crashes word prediction in 0.3.13 causes. Word prediction is scheduled to be completely reworked in 0.3.15/0.3.16.
> > If you experience any of the below listed bullet points or on [this issue collection](https://github.com/florisboard/florisboard/issues?q=is%3Aissue+is%3Aopen+label%3A%22area%3A+word-prediction%22), **DO NOT FILE A SEPARATE ISSUE**! These issues are already known and a separate issue only clutters the issue list even more. New word prediction issues will be closed as a duplicate of this issue here.
> > Word prediction in 0.3.13 and 0.3.14:
> > 
> > * Are English (US) only
> > * May not load even if enabled if the Java heap size is too small, in this case you just do not have word predictions atm
> > * If the dictionary loads it may cause various OOM (OutOfMemory) crashes, like in [OutOfMemory (OOM) Crash #677](https://github.com/florisboard/florisboard/issues/677). In this case disable word predictions and restart FlorisBoard
> > * Suggested words are always lowercase, the caps/shift key is ignored
> > * Suggested words do not suggest good words for some inputs
> > * ...
> 
> I know it's likely due to a bug but any idea why the suggestions in the smart bar would be words but literally with numbers at the end as I type for the current word? Thank you, sorry if I missed the answer somewhere.

It is always wise to read the [release notes](https://github.com/florisboard/florisboard/releases/tag/v0.3.14) upon installing a new version. This time they state:

> **Word suggestions**
> 
> This release temporary removes support for the completely buggy old word suggestions. If you still enable word suggestions the Smartbar would show a placeholder with the current word plus a number, to test out some UI layouts in the Smartbar. Please DO NOT file an issue regarding word suggestions not working, this is known and just clutters up the issue list unnecessarily. On a new word suggestion functionality will be worked on in the 0.4.0 development cycle.

**Comment by Glitchy-Tozier** [comment]
@patrickgold We can probably unpin this issue and close it, right?
If we want to warn users about limitations of spell-checking/word predictions in v0.4.0, we may want to use a new issue instead of the old OOM one.

**Comment by patrickgold** [comment]
Yep good idea. For the new word suggestions once I make the initial alpha release it will be accompanied by an announcement thread with the ability to give general feedback collected into one discussion. If users want to file specific bug reports/feature requests they should file a separate issue for that, but the details will be specified in that announcement thread once it is ready.
